common/djangoapps/student/views.py

```diff
@@ -2271,3 +2272,34 @@ def change_email_settings(request):
         track.views.server_track(request, "change-email-settings", {"receive_emails": "no", "course": course_id}, page='dashboard')
 
     return JsonResponse({"success": True})
+
+
+def _get_course_programs(user, user_enrolled_courses): # pylint: disable=invalid-name
+    """ Return a list of course key containing data required
+    for dashboard.
+
+    Given a user and an iterable of course keys, find all
+    the programs relevant to the user's dashboard and return them in a
+    dictionary keyed by the course_key.
+
+    Arguments:
+        user (user object): Currently logged-in User
+        course_keys (list): List of course keys in which user is enrolled
+
+    Returns:
+        Dictionary response containing programs or {}
+    """
+    course_programs = get_course_programs_for_dashboard(user, user_enrolled_courses)
+    programs_data = {}
+    for course_key, program in course_programs.items():
+        if program.get('status') == 'active' and program.get('category') == 'xseries':
+            programs_data[course_key] = {
```

Comment:
a program object missing any of the keys (course_codes, name, category, marketing_slug) would be unexpected and invalid.  So I'm worried about doing `.get(key, '')` on the dictionary because it might hide bugs, and the way such a program will end up displayed onscreen would be unpredictable.

Instead of doing `.get()` I suggest we directly access the expected keys (e.g. `program['course_codes']`) and catch KeyErrors at a higher level:

```
if program.get('status') == 'active' and program.get('category') == 'xseries':
    try:
        programs_data[course_key] = {
            'course_count': len(program['course_codes']),
            ...etc...
        }
    except KeyError:
        log.warn('program structure is invalid, skipping display: %r', program)
```
This way we avoid displaying anything half-baked to the user and log a warning in case there is a problem.